mkosi/__init__.py

```diff
@@ -739,15 +752,61 @@ def create_image(args: CommandLineArguments, root: str, for_cache: bool) -> Opti
         disable_cow(f.name)
         f.truncate(image_size(args))
 
-        table, run_sfdisk = determine_partition_table(args)
+        exec_sfdisk(args, f)
 
-        if run_sfdisk:
-            run(["sfdisk", "--color=never", f.name], input=table.encode("utf-8"))
-            run(["sync"])
+    return f
 
-        args.ran_sfdisk = run_sfdisk
 
-    return f
+def refresh_partition_table(args: CommandLineArguments, f: BinaryIO) -> None:
+    if not args.output_format.is_disk():
+        return None
+
+    # Let's refresh all UUIDs and labels to match the new build. This
+    # is called whenever we reuse a cached image, to ensure that the
+    # UUIDs/labels of partitions are generated the same way as for
+    # non-cached builds. Note that we refresh the UUIDs/labels simply
+    # by invoking sfdisk again. If the build parameters didn't change
+    # this should have the effect that offsets and sizes should remain
+    # identical, and we thus only update the UUIDs and labels.
+    #
+    # FIXME: One of those days we should generate the UUIDs as hashes
+    # of the used configuration, so that they remain stable as the
+    # configuration is identical.
+
+    with complete_step("Refreshing partition table", "Refreshed partition table") as output:
+        exec_sfdisk(args, f)
+
+
+def refresh_file_system(args: CommandLineArguments, dev: Optional[str], cached: bool) -> None:
+
+    if dev is None:
+        return
+    if not cached:
+        return
+
+    # Similar to refresh_partition_table() but refreshes the UUIDs of
+    # the file systems themselves. We want that build artifacts from
+    # cached builds are as similar as possible to those from uncached
+    # builds, and hence we want to randomize UUIDs explicitly like
+    # they are for uncached builds. This is particularly relevant for
+    # btrfs since it prohibits mounting multiple file systems at the
+    # same time that carry the same UUID.
+    #
+    # FIXME: One of those days we should generate the UUIDs as hashes
+    # of the used configuration, so that they remain stable as the
+    # configuration is identical.
+
+    with complete_step(f"Refreshing file system {dev}."):
+        if args.output_format == OutputFormat.gpt_btrfs:
+            run(["btrfstune", "-m", dev])
```

Comment:
The `-m` key is only available starting with kernel 5.0 and the current Debian stable is still on 4.19, when not using backports-kernels. I'm not too concerned about this, since Debian is missing repartd and all the other nice things, that would probably be needed, to actually make use of this.

An alternative would be to use `-M` and Python's `uuid` module.

Reply:
Changed it to use -M. PTAL